### PR TITLE
SOLR-17335: New 'vecSim' QParser

### PR DESCRIPTION
**Pull request by hossman**
https://issues.apache.org/jira/browse/SOLR-17335

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/search/neural/VecSimQParserPlugin.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+package org.apache.solr.search.neural;
+
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.search.QParser;
+import org.apache.solr.search.QParserPlugin;
+
+/** A neural query parser to run min-similarity search on Dense Vector fields. */
+public class VecSimQParserPlugin extends QParserPlugin {
```
Comment (alessandrobenedetti) [inline solr/core/src/java/org/apache/solr/search/neural/VecSimQParserPlugin.java:25]: same here, maybe "VectorSimilarityQParserPlugin"

File: solr/core/src/java/org/apache/solr/search/neural/VecSimQParserPlugin.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  public static final String NAME = "vecSim";
```
Comment (alessandrobenedetti) [inline solr/core/src/java/org/apache/solr/search/neural/VecSimQParserPlugin.java:26]: same here "vectorSimilarity"

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc [hunk 3]
```diff
@@ -237,14 +237,12 @@ client.add(Arrays.asList(d1, d2));
 --
 
 == Query Time
-This is the Apache Solr query approach designed to support dense vector search:
 
-=== knn Query Parser
-The `knn` k-nearest neighbors query parser allows to find the k-nearest documents to the target vector according to indexed dense vectors in the given field.  The set of documents can be Pre-Filtered to reduce the number of vector distance calculations that must be computed, and ensure the best `topK` are returned.
+Apache Solr provides two query parsers that work with dense vector fields, that each support differnet ways of matching documents based on vector similarity: The `knn` query parser, and the `vecSim` query parser.
```
Comment (alessandrobenedetti) [inline solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:241]: typo -> ", that each support different..."

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc [hunk 4]
```diff
@@ -237,14 +237,12 @@ client.add(Arrays.asList(d1, d2));
...
 
-The score for a retrieved document is the approximate distance to the target vector(defined by the similarityFunction configured at indexing time).
+Both parsers return scores for retrieved documents that is the approximate distance to the target vector (defined by the similarityFunction configured at indexing time) and both support "Pre-Filtering" the document graph to reduce the number of candidate vectors evaluated (with out needing to compute their vector similarity distances).
```
Comment (alessandrobenedetti) [inline solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:243]: typo? -> Both parsers return **a score** for retrieved documents that **is**...

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc [hunk 5]
```diff
@@ -293,22 +282,73 @@ Indicates that only `fq` filters with the specified `tag` should be considered f
...
+==== Explicit Pre-Filtering
+
+The `preFilter` parameter can be specified explicitly to reduce the number of candidate documents evaluated for the distance calculation:
 
 [source,text]
-?q={!knn f=vector topK=10 preFilter=inStock:true}[1.0, 2.0, 3.0, 4.0]
+?q={!vecSim f=vector minReturn=0.7 preFilter=inStock:true}[1.0, 2.0, 3.0, 4.0]
 
-In the above example, only documents matching the Pre-Filter `inStock:true` will be candidates for consideration when evaluating the k-nearest search against the specified vector.
+In the above example, only documents matching the Pre-Filter `inStock:true` will be candidates for consideration when evaluating the `knn` search against the specified vector.
```
Comment (alessandrobenedetti) [inline solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:351]: We mention here 'knn' search, but the example above is now a 'vectorSimilarity' query?

**Review by alessandrobenedetti — APPROVED** [review]
Thanks for the contribution! I just left some minor observations that in my opinion can improve the readability, but overall I don't have any main concerns with this pull request!
Well done and thanks for tackling this, it was on my TO-DO list to catch up with Lucene!

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc (hunk 4 above)
Comment (hossman) [reply solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:243]: Hmmm... I think the existing wording is more grammatically correct? ... your change implies that both of the parsers return a _single_ score for _all_ of the documents?

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc [hunk 6]
```diff
@@ -238,7 +238,7 @@ client.add(Arrays.asList(d1, d2));
 
 == Query Time
 
-Apache Solr provides two query parsers that work with dense vector fields, that each support differnet ways of matching documents based on vector similarity: The `knn` query parser, and the `vecSim` query parser.
+Apache Solr provides two query parsers that work with dense vector fields, that each support different ways of matching documents based on vector similarity: The `knn` query parser, and the `vecSim` query parser.
```
Comment (alessandrobenedetti) [inline solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:241]: vecSim -> vectorSimilarity

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc (hunk 4 above)
Comment (alessandrobenedetti) [reply solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:243]: I was just worried about the discordance between the number of 'scores' and the verb 'is' ?


Maybe : "Both parsers return **scores** for retrieved documents that **are** the approximate distance to the target vector" ?

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc [hunk 7]
```diff
@@ -348,7 +348,7 @@ The `preFilter` parameter can be specified explicitly to reduce the number of ca
 [source,text]
 ?q={!vecSim f=vector minReturn=0.7 preFilter=inStock:true}[1.0, 2.0, 3.0, 4.0]
```
Comment (alessandrobenedetti) [inline solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:349]: vecSim -> vectorSimilarity

**Review by alessandrobenedetti — CHANGES_REQUESTED** [review]
All looks great, aside from a couple of remaining points and renames that are needed in the documentation.

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc (hunk 6 above)
Comment (hossman) [reply solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:241]: Oh man ... i'm not sure how I managed to forget to do all the renames in the docs, i'll fix.

File: solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc (hunk 4 above)
Comment (hossman) [reply solr/solr-ref-guide/modules/query-guide/pages/dense-vector-search.adoc:243]: Ah, ok ... yeah sorry, i was focused on your switch to singular that i didn't notice that dissonance with "is" later in the sentence.

I'll fix.

**Pull request closed without merging**